frame/ibc/primitives/src/runtime_interface.rs

```diff
@@ -0,0 +1,194 @@
+#![allow(unused_imports)]
+
+use alloc::string::String;
+use base58::{FromBase58, ToBase58};
+use sp_core::{crypto::ByteArray, H256};
+use sp_runtime::AccountId32;
+use sp_runtime_interface::runtime_interface;
+use sp_std::prelude::*;
+
+#[derive(codec::Encode, codec::Decode)]
+pub enum SS58CodecError {
+	/// Invalid SS58 String
+	InvalidString,
+	/// Invalid Account id
+	InvalidAccountId,
+}
+
+pub fn blake2_256_verify_non_membership_proof(root: &H256, proof: &[Vec<u8>], key: &[u8]) -> bool {
+	sp_trie::verify_trie_proof::<sp_trie::LayoutV0<sp_runtime::traits::BlakeTwo256>, _, _, &[u8]>(
+		root,
+		proof,
+		&[(key, None)],
+	)
+	.is_ok()
+}
+
+pub fn sha2_512(message: &[u8]) -> [u8; 64] {
+	use sha2::Digest;
+	let mut hasher = sha2::Sha512::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 64];
+	res.copy_from_slice(&hash);
+	res
+}
+
+pub fn sha2_512_truncated(message: &[u8]) -> [u8; 32] {
+	use sha2::Digest;
+	let mut hasher = sha2::Sha512::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 32];
+	res.copy_from_slice(&hash[..32]);
+	res
+}
+
+pub fn sha3_512(message: &[u8]) -> [u8; 64] {
+	use sha3::Digest;
+	let mut hasher = sha3::Sha3_512::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 64];
+	res.copy_from_slice(&hash);
+	res
+}
+
+pub fn ripemd160(message: &[u8]) -> [u8; 20] {
+	use ripemd::Digest;
+	let mut hasher = ripemd::Ripemd160::new();
+	hasher.update(message);
+	let hash = hasher.finalize();
+	let mut res = [0u8; 20];
+	res.copy_from_slice(&hash);
+	res
+}
+
+pub fn ss58_to_account_id_32(raw_str: &str) -> Result<[u8; 32], SS58CodecError> {
```

Comment:
what's `ss`?

Reply:
its ss58

Reply:
ok, just for the ones (like me) that know nothing about this: https://wiki.polkadot.network/docs/build-ss58-registry